### DISCUSSION TITLE
Penalties for being late

### DISCUSSION THREAD

**Discussion by odashevsky** [google/or-tools#3841] [Routing (and legacy CP) questions]
Hi,

We're solving a VRP problem where vehicles pickup and drop off load at locations within certain time windows. There's a window where a load could be picked up at location A and window the load could be dropped-off at location B. The drop-off is actually an appointment. Appointment is at time T. Its OK to be early at the appointment T1 and its OK to be late at the appointment T2. So, the drop-off window is T1 - T2 where T is between T1 and T2. However, the algorithm needs to make everything possible to make it between T1 and T, its OK to be as late as T2 (but never beyond T2). 
Any suggestions on how to accomplish this?

**Comment by jmarca** [comment]
Can you be a little more clear?  I’m confused whether the time windows affect the pickup at A *and* the drop off at B, or both. And why doesn’t a regular time window work?JamesOn Jul 2, 2023, at 16:06, Oleg Dashevsky ***@***.***> wrote:﻿
Hi,
We're solving a VRP problem where vehicles pickup and drop off load at locations within certain time windows. There's a window where a load could be picked up at location A and window the load could be dropped-off at location B. The drop-off is actually an appointment. Appointment is at time T. Its OK to be early at the appointment T1 and its OK to be late at the appointment T2. So, the drop-off window is T1 - T2 where T is between T1 and T2. However, the algorithm needs to make everything possible to make it between T1 and T, its OK to be as late as T2 (but never beyond T2).
Any suggestions on how to accomplish this?

**Reply by odashevsky** [reply]
No, pickup at A is unaffected for sake of the question. The requirement is that drop-off has to happen between T1 and T2 and appointment time T is between T1 and T2, but the solver should try to avoid dropping-off between T and T2. It is allowed, but it should avoid it and find solutions where drop-off can happen between T1 and T and only if there's no good alternative drop-off between T and T2. Hope this makes sense. Basically the requirement is "try to avoid being late to appointments".

**Comment by jmarca** [comment]
One idea (but I haven’t given it much thought) is to set the time window as usual (and of course confirm that it works as expected) then use a soft upper bound (https://developers.google.com/optimization/reference/python/constraint_solver/pywrapcp#setcumulvarsoftupperbound) to encourage hitting at most your target time. 

James

**Reply by odashevsky** [reply]
Thanks, James! That helped.
